### ISSUE TITLE
Combobox v2 issue with dynamic fetched options in nextjs

### ISSUE THREAD

**Issue by lasseklovstad** [tailwindlabs/headlessui#3184]
**What package within Headless UI are you using?**

@headlessui/react

**What version of that package are you using?**

v2.0.3

**What browser are you using?**

Chrome

**Reproduction URL**

[Reproduction CodeSandbox](https://codesandbox.io/p/devbox/elated-wave-xq2cft?file=%2Fapp%2Fpage.tsx%3A10%2C21&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clv6n5fs70007356n59sz4niz%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clv6n5fs70002356nhvqk5e3j%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clv6n5fs70004356ny1o6qsmb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clv6n5fs70006356nz7717e0i%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clv6n5fs70002356nhvqk5e3j%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv6n5fs70001356ndoricerp%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clvxop8hz0002356o6k7pkqwl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A10%252C%2522startColumn%2522%253A21%252C%2522endLineNumber%2522%253A10%252C%2522endColumn%2522%253A21%257D%255D%252C%2522filepath%2522%253A%2522%252Fapp%252Fpage.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clv6n5fs70002356nhvqk5e3j%2522%252C%2522activeTabId%2522%253A%2522clvxop8hz0002356o6k7pkqwl%2522%257D%252C%2522clv6n5fs70006356nz7717e0i%2522%253A%257B%2522id%2522%253A%2522clv6n5fs70006356nz7717e0i%2522%252C%2522activeTabId%2522%253A%2522clvxob4t40085356ozb2888ct%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv6n5fs70005356no5b0ioxj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clvxob4t40085356ozb2888ct%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clv6n5fs70004356ny1o6qsmb%2522%253A%257B%2522id%2522%253A%2522clv6n5fs70004356ny1o6qsmb%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv6n5fs70003356ns9l86x9o%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clv6n5fs70003356ns9l86x9o%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Describe your issue**
![combox-issue](https://github.com/tailwindlabs/headlessui/assets/43168787/705a43dd-9f2f-4fbd-8d05-9c475dbacf9a)

After selecting an option and changing the input value the Input value resets back to the selected option. This is not how it was in the latest v1. This issue also seems only to occure in Nextjs. I could not reproduce it with the vite issue template.

1. Type the letter "T" in inputfield and select the "Tailwind" option
2. After selecting we want to search for something new, so we backspace 4 letters.
3. This will trigger react query to run again and the input is reset to the selected value.

The excpected behavior is to not change the displayValue when changing the inputValue.

**Comment by juliancoleman** [comment]
Here to say I'm experiencing the same thing

**Comment by juliancoleman** [comment]
@lasseklovstad Following up here: I can temporarily get around this by setting the Combobox value to `null` if the input value has changed. This obviously isn't great and impacts the UX negatively (e.g. blurring the field will not reset to the previously selected value by default), which means more code will need to be written by you to restore the ideal UX.

Here's some (simplified) code I'm using that's running in production now using `@headlessui/react ^2.0.3`

```tsx
type Place = {
  code: string; // can be empty string
  type: string; // can be empty string
};

export type LocationPickerProps = {
  /**
   * Available options to populate the menu
   */
  options: LocationResponse;
  /**
   * The selected value
   */
  value: Place | null;
  /**
   * Callback whenever a new selection is made
   */
  onChange(value: Place | null): void;
  /**
   * Callback whenever the search input value changes
   */
  onSearch(search: string): void;
};

export function LocationPicker(props: LocationPickerProps) {
  const {
    onChange,
    onSearch,
    options,
    value,
  } = props;

  const handleSearch = useMemo(
    () => debounce((input: string) => onSearch(input)),
    [onSearch],
  );

return (
    <Combobox
      value={value}
      onChange={(value) => onChange(value)}
      as="div"
      className="w-full"
      by="code"
      immediate
    >
      <Label className="flex w-full items-center gap-2.5 rounded-lg border border-grey-200 bg-grey-50 px-3 py-2.5 transition-all focus-within:border-grey-300 hover:border-grey-400">
        <ComboboxInput
          autoComplete="off"
          placeholder="Choose location"
          onChange={({ target }) => {
            onChange(null); // <-- this fixes the bug in question
            handleSearch(target.value);
          }}
          displayValue={(location: Place | null) => {
            if (
              location == null ||
              (location.type === "" && location.code === "")
            ) {
              return "";
            }

            const found = options.find(
              (o) =>
                o.code === location.code &&
                o.granularLocationType === location.type,
            );

            return found?.title ?? "";
          }}
          className="flex-1 bg-transparent text-sm outline-none"
        />
      </Label>

      {/* ... */}

    </Combobox>
  );
}
```

**Comment by RobinMalfait** [comment]
This should be fixed by #3259, and will be available in the next release.

You can already try it using:
- `npm install @headlessui/react@insiders`.

**Comment by juliancoleman** [comment]
Thanks @RobinMalfait 👍
